src/history/file.rs

```diff
@@ -234,320 +236,3 @@ fn read_zero_padded(file: &mut File, mut dest: &mut [u8]) -> std::io::Result<()>
     dest.fill(0u8);
     Ok(())
 }
-
-fn replace_all(s: &mut Vec<u8>, needle: &[u8], replacement: &[u8]) {
-    let mut offset = 0;
-    while let Some(relpos) = subslice_position(&s[offset..], needle) {
-        offset += relpos;
-        s.splice(offset..(offset + needle.len()), replacement.iter().copied());
-        offset += replacement.len();
-    }
-}
-
-/// Support for escaping and unescaping the nonstandard "yaml" format introduced in fish 2.0.
-fn escape_yaml_fish_2_0(s: &mut Vec<u8>) {
-    replace_all(s, b"\\", b"\\\\"); // replace one backslash with two
-    replace_all(s, b"\n", b"\\n"); // replace newline with backslash + literal n
-}
-
-#[inline(always)]
-/// Unescapes the fish-specific yaml variant, if it requires it.
-fn maybe_unescape_yaml_fish_2_0(s: &[u8]) -> Cow<[u8]> {
-    // This is faster than s.contains(b'\\') and can be auto-vectorized to SIMD. See benchmark note
-    // on unescape_yaml_fish_2_0().
-    if !s.iter().copied().fold(false, |acc, b| acc | (b == b'\\')) {
-        return s.into();
-    }
-    unescape_yaml_fish_2_0(s).into()
-}
-
-/// Unescapes the fish-specific yaml variant. Use [`maybe_unescape_yaml_fish_2_0()`] if you're not
-/// positive the input contains an escape.
-
-// This function is called on every input event and shows up heavily in all flamegraphs.
-// Various approaches were benchmarked against real-world fish-history files on lines with escapes,
-// and this implementation (chunk_loop_box) won out. Make changes with care!
-//
-// Benchmarks and code: https://github.com/mqudsi/fish-yaml-unescape-benchmark
-fn unescape_yaml_fish_2_0(s: &[u8]) -> Vec<u8> {
-    // Safety: we never read from this box; we only write to it then read what we've written.
-    // Rationale: This function is in a very hot loop and this benchmarks around 8% faster on
-    // real-world escaped yaml samples from the fish history file.
-    //
-    // This is a very long way around of writing `Box::new_uninit_slice(s.len())`, which
-    // requires the unstablized nightly-only feature new_unit (#63291). It optimizes away.
-    let mut result: Box<[u8]> = std::iter::repeat_with(std::mem::MaybeUninit::uninit)
-        .take(s.len())
-        .map(|b| unsafe { b.assume_init() })
```

Comment:
Preexisting, but this is 100% UB. The safety comment doesn't actually address the safety requirements of `assume_init()`, which are totally unrelated to whether or not the value is "read". Needs fixing.

cc @mqudsi

Reply:
@Xiretza https://github.com/fish-shell/fish-shell/commit/6c37103b7c4715e47f9ab5492c1db0dd8f8e19d8

Reply:
It seemed fine to me, I don't think it's UB if it never happens.

Reply:
What do you mean by "it never happens"? This called `MaybeUninit::uninit().assume_init()` `s.len()` times.

Reply:
The values are never read, so the values don't matter. It's true that we are cheating/lying to `MaybeUninit` but that's fine because the destructor of `u8` is trivial.

Reply:
No, it's not. The documentation for `MaybeUninit` *specifically* calls out that `MaybeUninit::<i32>::uninit().assume_init()` is UB, the same is true for `u8`.

Reply:
My view here: I think the concern is regarding the potential for UB, i.e. code paths that the compiler cannot guarantee are safe. The code in it's current state does not result in undefined behavior, as the values are never read. However it does feel fragile to me, as this code could change in the future, and it's not immediately obvious to me that reading at a certain point could be unsafe.

Reply:
Okay you're right we were UB according to the terms of `MaybeUninit`. There are no practical consequences to violating this contract but yeah it should be written differently